### PR TITLE
Editor: Added a new panel for editing key bindings

### PR DESCRIPTION
**Pull request by TyLindberg**
By adding this panel, users now have the ability to edit key bindings for the translate, rotate, scale, and undo buttons. This could be useful for those with different keyboard configurations and also allows more key bindings to be easily added without cluttering index.html. Currently, I called the panel 'CONTROLS' and didn't allow for the binding of modifier keys such as Command/Control.

Here is an image of what the panel looks like:
screenshot-2017-11-9 three js editor

There are a bunch of changes that can be made with this, so definitely let me know your thoughts. Hopefully we can leverage this to add more commands to the editor in the future.

### CODE REVIEW

**Comment by mrdoob** [comment]
Looking good!

File: editor/js/Config.js [hunk 1]
```diff
@@ -19,7 +19,12 @@ var Config = function ( name ) {
 
 		'project/vr': false,
 
-		'settings/history': false
+		'settings/history': false,
+
+		'controls/translate': 'w',
```
Comment (mrdoob) [inline editor/js/Config.js]: Do you mind renaming these to `settings/controls/translate`...?

File: editor/js/Sidebar.Controls.js [hunk 2]
```diff
@@ -0,0 +1,154 @@
...
+
+				break;
+
+			case editor.config.getKey( 'controls/scale' ): // Scaling transform mode
+
+				editor.signals.transformModeChanged.dispatch( 'scale' );
+
+				break;
+
+			case editor.config.getKey( 'controls/undo' ).toLowerCase(): // Register Ctrl/Command-Z for Undo and Ctrl/Command-Shift-Z for Redo
```
Comment (mrdoob) [inline editor/js/Sidebar.Controls.js]: I bet having caps locked messes with this...

Comment (TyLindberg) [reply editor/js/Sidebar.Controls.js]: Ahh good call, I haven't tested how caps lock affects the controls. There is currently weird behavior where pressing CTRL+SHIFT+Z results in the key being 'Z' on Chrome and 'z' on Safari and Firefox hence why I'm checking for both upper and lower case for undo/redo. I'll test out caps lock and let you know what behavior I see.

File: editor/js/Config.js (hunk 1 above)
Comment (TyLindberg) [reply editor/js/Config.js]: Done!

**Comment by TyLindberg** [comment]
So with the current implementation, having caps lock pressed will result in the translate/rotate/scale controls not being registered. This is because those controls are case sensitive. Undo/redo is not case sensitive, so it is unaffected by having caps lock pressed, and in addition, you still need to hold shift for a redo to be triggered instead of undo.

This is mostly a design decision whether we want to keep it like this or not. By allowing things like translate/rotate/scale to be case sensitive, it lets users bind those commands to SHIFT+<key> without much extra work. However, we could make bindings case insensitive if we felt that was more intuitive or a more consistent experience.

**Comment by mrdoob** [comment]
Hmm, okay. I'll merge and experiment with this.

**Comment by mrdoob** [comment]
Thanks!

**Comment by mrdoob** [comment]
Ended up making it case-insensitive (forced to lowercase).

https://github.com/mrdoob/three.js/commit/9dbe8d0cf1bf26f912e877c34bca1cf088987cb8

**Comment by mrdoob** [comment]
Also, this makes it work in Safari.

https://github.com/mrdoob/three.js/commit/d35ce0178f031f15e45116f36319cfb92cbb36d5

**Comment by TyLindberg** [comment]
Ah, perfect! I'm glad you were able to find a way to make that consistent between browsers.

And forcing to lowercase seems pretty smart to me. In the future it may make sense to add the ability to bind certain modifier keys such as shift or alt. However, keeping everything lowercase for now seems like it will cause the least issues.

Thanks for merging that in!

**Comment by mrdoob** [comment]
🙌

**Pull request merged**